File: src/precalculus_trig_identities_hard.py
"""
precalculus - trig_identities (hard)
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 6)

    if problem_type == 1:
        # Express sin(A)*cos(B) products as sum/difference (product-to-sum)
        # With wide variety of angle multiples
        variant = randint(1, 4)
        a_vals = [1, 2, 3, 4, 5]
        b_vals = [1, 2, 3, 4, 5]
        a = choice(a_vals)
        b = choice(b_vals)
        while a == b:
            b = choice(b_vals)

        if variant == 1:
            # 2 sin(ax) cos(bx) = sin((a+b)x) + sin((a-b)x)
            sum_c = a + b
            diff_c = a - b
            ans = sin(sum_c * x) + sin(diff_c * x)
            return problem(
                question=f"Express $2\\sin({a}x)\\cos({b}x)$ as a sum/difference of trig functions.",
                answer=simplify(ans),
                difficulty=(1600, 1750),
                topic="precalculus/trig_identities",
                solution=steps(
                    "Product-to-sum: $2\\sin A\\cos B = \\sin(A+B) + \\sin(A-B)$",
                    f"$A = {a}x$, $B = {b}x$",
                    f"$= \\sin({sum_c}x) + \\sin({diff_c}x)$"
                ),
                grading_mode="equivalent"
            )
        elif variant == 2:
            # 2 cos(ax) cos(bx) = cos((a-b)x) + cos((a+b)x)
            ans = cos((a - b) * x) + cos((a + b) * x)
            return problem(
                question=f"Express $2\\cos({a}x)\\cos({b}x)$ as a sum.",
                answer=simplify(ans),
                difficulty=(1600, 1750),
                topic="precalculus/trig_identities",
                solution=steps(
                    "Product-to-sum: $2\\cos A\\cos B = \\cos(A-B) + \\cos(A+B)$",
                    f"$= \\cos({abs(a-b)}x) + \\cos({a+b}x)$"
                ),
                grading_mode="equivalent"
            )
        elif variant == 3:
            # 2 sin(ax) sin(bx) = cos((a-b)x) - cos((a+b)x)
            ans = cos((a - b) * x) - cos((a + b) * x)
            return problem(
                question=f"Express $2\\sin({a}x)\\sin({b}x)$ as a difference.",
                answer=simplify(ans),
                difficulty=(1600, 1750),
                topic="precalculus/trig_identities",
                solution=steps(
                    "Product-to-sum: $2\\sin A\\sin B = \\cos(A-B) - \\cos(A+B)$",
                    f"$= \\cos({abs(a-b)}x) - \\cos({a+b}x)$"
                ),
                grading_mode="equivalent"
            )
        else:
            # Verify: 2sin(3x)cos(x) = sin(4x) + sin(2x), with random a
            c1 = choice([2, 3, 4, 5])
            c2 = choice([1, 2, 3])
            sum_c = c1 + c2
            diff_c = c1 - c2
            ans = sin(sum_c * x) + sin(diff_c * x)
            return problem(
                question=f"Verify: $2\\sin({c1}x)\\cos({c2}x) = \\sin({sum_c}x) + \\sin({diff_c}x)$",
                answer=True,
                difficulty=(1700, 1850),
                topic="precalculus/trig_identities",
                solution=steps(
                    "Use product-to-sum: $2\\sin A\\cos B = \\sin(A+B) + \\sin(A-B)$",
                    f"With $A={c1}x$, $B={c2}x$: $\\sin({sum_c}x) + \\sin({diff_c}x)$ ✓"
                ),
                answer_type="boolean"
            )

    elif problem_type == 2:
        # Sum-to-product identities
        variant = randint(1, 3)
        a = choice([2, 3, 4, 5])
        b = choice([1, 2, 3])
        while a == b:
            b = choice([1, 2, 3])
        half_sum = Rational(a + b, 2)
        half_diff = Rational(a - b, 2)

        if variant == 1:
            # sin(ax) + sin(bx) = 2 sin((a+b)/2 * x) cos((a-b)/2 * x)
            lhs_str = f"\\sin({a}x) + \\sin({b}x)"
            ans = 2 * sin(half_sum * x) * cos(half_diff * x)
            return problem(
                question=f"Express ${lhs_str}$ as a product.",
                answer=simplify(ans),
                difficulty=(1650, 1800),
                topic="precalculus/trig_identities",
                solution=steps(
                    "Sum-to-product: $\\sin A + \\sin B = 2\\sin\\frac{A+B}{2}\\cos\\frac{A-B}{2}$",
                    f"$= 2\\sin({latex(half_sum)}x)\\cos({latex(half_diff)}x)$"
                ),
                grading_mode="equivalent"
            )
        elif variant == 2:
            # sin(ax) - sin(bx) = 2 cos((a+b)/2 * x) sin((a-b)/2 * x)
            lhs_str = f"\\sin({a}x) - \\sin({b}x)"
            ans = 2 * cos(half_sum * x) * sin(half_diff * x)
            return problem(
                question=f"Express ${lhs_str}$ as a product.",
                answer=simplify(ans),
                difficulty=(1650, 1800),
                topic="precalculus/trig_identities",
                solution=steps(
                    "Sum-to-product: $\\sin A - \\sin B = 2\\cos\\frac{A+B}{2}\\sin\\frac{A-B}{2}$",
                    f"$= 2\\cos({latex(half_sum)}x)\\sin({latex(half_diff)}x)$"
                ),
                grading_mode="equivalent"
            )
        else:
            # cos(ax) + cos(bx)
            lhs_str = f"\\cos({a}x) + \\cos({b}x)"
            ans = 2 * cos(half_sum * x) * cos(half_diff * x)
            return problem(
                question=f"Express ${lhs_str}$ as a product.",
                answer=simplify(ans),
                difficulty=(1650, 1800),
                topic="precalculus/trig_identities",
                solution=steps(
                    "Sum-to-product: $\\cos A + \\cos B = 2\\cos\\frac{A+B}{2}\\cos\\frac{A-B}{2}$",
                    f"$= 2\\cos({latex(half_sum)}x)\\cos({latex(half_diff)}x)$"
                ),
                grading_mode="equivalent"
            )

    elif problem_type == 3:
        # Solve trig equation using identities on [0, 2pi)
        variant = randint(1, 4)

        if variant == 1:
            # cos(2x) = sin(x)
            ans = FiniteSet(pi/6, 5*pi/6, 3*pi/2)
            return problem(
                question="Solve $\\cos(2x) = \\sin(x)$ for $x \\in [0, 2\\pi)$",
                answer=ans,
                difficulty=(1700, 1850),
                topic="precalculus/trig_identities",
                solution=steps(
                    "Use $\\cos(2x) = 1 - 2\\sin^2(x)$",
                    "$1 - 2\\sin^2 x = \\sin x \\Rightarrow 2\\sin^2 x + \\sin x - 1 = 0$",
                    "$(2\\sin x - 1)(\\sin x + 1) = 0$",
                    "$\\sin x = \\frac{1}{2}$ or $\\sin x = -1$",
                    "$x = \\frac{\\pi}{6}, \\frac{5\\pi}{6}, \\frac{3\\pi}{2}$"
                ),
                answer_type="set"
            )
        elif variant == 2:
            # sin(2x) = cos(x)
            ans = FiniteSet(pi/6, pi/2, 5*pi/6, 3*pi/2)
            return problem(
                question="Solve $\\sin(2x) = \\cos(x)$ for $x \\in [0, 2\\pi)$",
                answer=ans,
                difficulty=(1700, 1850),
                topic="precalculus/trig_identities",
                solution=steps(
                    "Use $\\sin(2x) = 2\\sin x\\cos x$",
                    "$2\\sin x\\cos x - \\cos x = 0 \\Rightarrow \\cos x(2\\sin x - 1) = 0$",
                    "$\\cos x = 0$ or $\\sin x = \\frac{1}{2}$",
                    "$x = \\frac{\\pi}{2}, \\frac{3\\pi}{2}, \\frac{\\pi}{6}, \\frac{5\\pi}{6}$"
                ),
                answer_type="set"
            )
        elif variant == 3:
            # 2 sin^2(x) - sin(x) - 1 = 0
            ans = FiniteSet(7*pi/6, 3*pi/2, 11*pi/6, pi/2)
            # 2sin^2 - sin - 1 = 0 => (2sin+1)(sin-1)=0 => sin=-1/2 or sin=1
            # sin=-1/2: x=7pi/6, 11pi/6; sin=1: x=pi/2
            ans = FiniteSet(pi/2, 7*pi/6, 11*pi/6)
            return problem(
                question="Solve $2\\sin^2(x) - \\sin(x) - 1 = 0$ for $x \\in [0, 2\\pi)$",
                answer=ans,
                difficulty=(1700, 1850),
                topic="precalculus/trig_identities",
                solution=steps(
                    "Factor: $(2\\sin x + 1)(\\sin x - 1) = 0$",
                    "$\\sin x = -\\frac{1}{2}$ or $\\sin x = 1$",
                    "$x = \\frac{7\\pi}{6}, \\frac{11\\pi}{6}$ (for $-\\frac{1}{2}$); $x = \\frac{\\pi}{2}$ (for $1$)"
                ),
                answer_type="set"
            )
        else:
            # cos(2x) + cos(x) = 0
            # 2cos^2(x)-1+cos(x)=0 => (2cos(x)-1)(cos(x)+1)=0
            # cos=1/2: pi/3, 5pi/3; cos=-1: pi
            ans = FiniteSet(pi/3, pi, 5*pi/3)
            return problem(
                question="Solve $\\cos(2x) + \\cos(x) = 0$ for $x \\in [0, 2\\pi)$",
                answer=ans,
                difficulty=(1700, 1850),
                topic="precalculus/trig_identities",
                solution=steps(
                    "Use $\\cos(2x) = 2\\cos^2 x - 1$",
                    "$2\\cos^2 x - 1 + \\cos x = 0 \\Rightarrow (2\\cos x - 1)(\\cos x + 1) = 0$",
                    "$\\cos x = \\frac{1}{2}$ or $\\cos x = -1$",
                    "$x = \\frac{\\pi}{3}, \\frac{5\\pi}{3}, \\pi$"
                ),
                answer_type="set"
            )

    elif problem_type == 4:
        # Express as single trig function using sum/difference: sin(ax+bx), cos, etc.
        variant = randint(1, 3)
        a = choice([2, 3, 4, 5, 6])
        b = choice([1, 2, 3, 4])
        while a == b:
            b = choice([1, 2, 3, 4])

        if variant == 1:
            ans = sin((a + b) * x)
            return problem(
                question=f"Express $\\sin({a}x)\\cos({b}x) + \\cos({a}x)\\sin({b}x)$ as a single trig function.",
                answer=ans,
                difficulty=(1600, 1700),
                topic="precalculus/trig_identities",
                solution=steps(
                    "Sine addition formula: $\\sin(A+B) = \\sin A\\cos B + \\cos A\\sin B$",
                    f"$A = {a}x$, $B = {b}x$: $\\sin({a+b}x)$"
                ),
                grading_mode="equivalent"
            )
        elif variant == 2:
            ans = cos((a + b) * x)
            return problem(
                question=f"Express $\\cos({a}x)\\cos({b}x) - \\sin({a}x)\\sin({b}x)$ as a single trig function.",
                answer=ans,
                difficulty=(1600, 1700),
                topic="precalculus/trig_identities",
                solution=steps(
                    "Cosine addition formula: $\\cos(A+B) = \\cos A\\cos B - \\sin A\\sin B$",
                    f"$A = {a}x$, $B = {b}x$: $\\cos({a+b}x)$"
                ),
                grading_mode="equivalent"
            )
        else:
            ans = sin((a - b) * x)
            return problem(
                question=f"Express $\\sin({a}x)\\cos({b}x) - \\cos({a}x)\\sin({b}x)$ as a single trig function.",
                answer=ans,
                difficulty=(1600, 1700),
                topic="precalculus/trig_identities",
                solution=steps(
                    "Sine difference formula: $\\sin(A-B) = \\sin A\\cos B - \\cos A\\sin B$",
                    f"$= \\sin({a-b}x)$"
                ),
                grading_mode="equivalent"
            )

    elif problem_type == 5:
        # Half-angle and power reduction, varied n
        variant = randint(1, 4)
        n = choice([2, 3, 4, 5, 6])

        if variant == 1:
            ans = (1 - cos(2 * n * x)) / 2
            return problem(
                question=f"Express $\\sin^2({n}x)$ without any squared trig functions.",
                answer=ans,
                difficulty=(1650, 1800),
                topic="precalculus/trig_identities",
                solution=steps(
                    f"Power reduction: $\\sin^2\\theta = \\dfrac{{1-\\cos(2\\theta)}}{{2}}$",
                    f"$\\sin^2({n}x) = \\dfrac{{1 - \\cos({2*n}x)}}{{2}}$"
                ),
                grading_mode="equivalent"
            )
        elif variant == 2:
            ans = (1 + cos(2 * n * x)) / 2
            return problem(
                question=f"Express $\\cos^2({n}x)$ without any squared trig functions.",
                answer=ans,
                difficulty=(1650, 1800),
                topic="precalculus/trig_identities",
                solution=steps(
                    f"Power reduction: $\\cos^2\\theta = \\dfrac{{1+\\cos(2\\theta)}}{{2}}$",
                    f"$\\cos^2({n}x) = \\dfrac{{1 + \\cos({2*n}x)}}{{2}}$"
                ),
                grading_mode="equivalent"
            )
        elif variant == 3:
            # sin^4(x) in terms of cos(2x), cos(4x)
            # sin^4 = (sin^2)^2 = ((1-cos2x)/2)^2 = (1-2cos2x+cos^2(2x))/4
            # = (1 - 2cos2x + (1+cos4x)/2)/4 = (3/8 - cos2x/2 + cos4x/8)
            ans = Rational(3, 8) - cos(2 * x) / 2 + cos(4 * x) / 8
            return problem(
                question="Express $\\sin^4(x)$ in terms of $\\cos(2x)$ and $\\cos(4x)$ (no powers).",
                answer=ans,
                difficulty=(1750, 1900),
                topic="precalculus/trig_identities",
                solution=steps(
                    "$\\sin^4 x = (\\sin^2 x)^2 = \\left(\\dfrac{1-\\cos 2x}{2}\\right)^2$",
                    "$= \\dfrac{1 - 2\\cos 2x + \\cos^2 2x}{4}$",
                    "$\\cos^2 2x = \\dfrac{1+\\cos 4x}{2}$",
                    "$= \\dfrac{1 - 2\\cos 2x + \\frac{1+\\cos 4x}{2}}{4} = \\dfrac{3}{8} - \\dfrac{\\cos 2x}{2} + \\dfrac{\\cos 4x}{8}$"
                ),
                grading_mode="equivalent"
            )
        else:
            # half-angle: sin(x/2)^2 = (1-cos x)/2
            return problem(
                question="Use the half-angle formula to simplify $\\sin^2\\!\\left(\\dfrac{x}{2}\\right)$.",
                answer=(1 - cos(x)) / 2,
                difficulty=(1650, 1800),
                topic="precalculus/trig_identities",
                solution=steps(
                    "$\\sin^2\\!\\left(\\dfrac{\\theta}{2}\\right) = \\dfrac{1 - \\cos\\theta}{2}$",
                    "Substitute $\\theta = x$: $\\dfrac{1 - \\cos x}{2}$"
                ),
                grading_mode="equivalent"
            )

    else:
        # Complex simplifications: (1-sin/cos+cos/(1-sin)) = 2 sec, etc.
        variant = randint(1, 4)

        if variant == 1:
            # sin^2/(1-cos) = 1 + cos
            ans = 1 + cos(x)
            return problem(
                question="Simplify $\\dfrac{\\sin^2 x}{1 - \\cos x}$",
                answer=ans,
                difficulty=(1650, 1800),
                topic="precalculus/trig_identities",
                solution=steps(
                    "Use $\\sin^2 x = (1-\\cos x)(1+\\cos x)$",
                    "$\\dfrac{(1-\\cos x)(1+\\cos x)}{1-\\cos x} = 1 + \\cos x$"
                ),
                grading_mode="equivalent"
            )
        elif variant == 2:
            # (1-sin x)/cos x + cos x/(1-sin x) = 2 sec x
            ans = 2 / cos(x)
            return problem(
                question="Simplify $\\dfrac{1-\\sin x}{\\cos x} + \\dfrac{\\cos x}{1-\\sin x}$",
                answer=ans,
                difficulty=(1750, 1950),
                topic="precalculus/trig_identities",
                solution=steps(
                    "Common denominator $\\cos x(1-\\sin x)$:",
                    "$\\dfrac{(1-\\sin x)^2 + \\cos^2 x}{\\cos x(1-\\sin x)}$",
                    "Numerator: $1 - 2\\sin x + \\sin^2 x + \\cos^2 x = 2 - 2\\sin x = 2(1-\\sin x)$",
                    "$= \\dfrac{2(1-\\sin x)}{\\cos x(1-\\sin x)} = \\dfrac{2}{\\cos x} = 2\\sec x$"
                ),
                grading_mode="equivalent"
            )
        elif variant == 3:
            # sec x / (sec x - tan x) - sec x / (sec x + tan x) = 2 tan x sec x
            ans = 2 * tan(x) / cos(x)
            return problem(
                question="Simplify $\\dfrac{\\sec x}{\\sec x - \\tan x} - \\dfrac{\\sec x}{\\sec x + \\tan x}$",
                answer=ans,
                difficulty=(1750, 1950),
                topic="precalculus/trig_identities",
                solution=steps(
                    "Common denominator: $(\\sec x - \\tan x)(\\sec x + \\tan x) = \\sec^2 x - \\tan^2 x = 1$",
                    "$\\sec x(\\sec x + \\tan x) - \\sec x(\\sec x - \\tan x) = 2\\sec x\\tan x$",
                    "Answer: $2\\sec x\\tan x$"
                ),
                grading_mode="equivalent"
            )
        else:
            # (tan x + cot x)^2 = sec^2 x + csc^2 x
            ans = sec(x)**2 + csc(x)**2
            return problem(
                question="Simplify $(\\tan x + \\cot x)^2$",
                answer=ans,
                difficulty=(1700, 1850),
                topic="precalculus/trig_identities",
                solution=steps(
                    "$\\tan x + \\cot x = \\dfrac{\\sin x}{\\cos x} + \\dfrac{\\cos x}{\\sin x} = \\dfrac{\\sin^2 x+\\cos^2 x}{\\sin x\\cos x} = \\dfrac{1}{\\sin x\\cos x}$",
                    "$(\\tan x + \\cot x)^2 = \\dfrac{1}{\\sin^2 x\\cos^2 x} = \\sec^2 x\\csc^2 x$",
                    "Also equals $\\sec^2 x + \\csc^2 x$ (expanding $(\\tan+\\cot)^2$ directly)"
                ),
                grading_mode="equivalent"
            )

emit(generate())

File: src/precalculus_trig_identities_hard.jl
# precalculus - trig_identities (hard)
# Generated: 2026-03-08T20:29:25.200467

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("precalculus/trig_identities")
    
    problem_type = rand(1:5)
    
    if problem_type == 1
        # Prove a double angle identity with specific angle
        angle_mult = choice([2, 3, 4])
        base_angle = choice([15, 18, 22, 30, 36, 45])
        
        # Choose identity type
        identity_choice = rand(1:3)
        
        if identity_choice == 1
            # sin(2θ) = 2sin(θ)cos(θ)
            lhs_expr = "\\sin($(angle_mult * base_angle)^\\circ)"
            if angle_mult == 2
                rhs_expr = "2\\sin($(base_angle)^\\circ)\\cos($(base_angle)^\\circ)"
            else
                rhs_expr = "\\sin(2 \\cdot $(div(angle_mult * base_angle, 2))^\\circ)"
            end
        elseif identity_choice == 2
            # cos(2θ) = cos²(θ) - sin²(θ)
            lhs_expr = "\\cos($(angle_mult * base_angle)^\\circ)"
            if angle_mult == 2
                rhs_expr = "\\cos^2($(base_angle)^\\circ) - \\sin^2($(base_angle)^\\circ)"
            else
                rhs_expr = "\\cos(2 \\cdot $(div(angle_mult * base_angle, 2))^\\circ)"
            end
        else
            # tan(2θ) = 2tan(θ)/(1-tan²(θ))
            lhs_expr = "\\tan($(angle_mult * base_angle)^\\circ)"
            if angle_mult == 2
                rhs_expr = "\\frac{2\\tan($(base_angle)^\\circ)}{1 - \\tan^2($(base_angle)^\\circ)}"
            else
                rhs_expr = "\\tan(2 \\cdot $(div(angle_mult * base_angle, 2))^\\circ)"
            end
        end
        
        problem(
            question="Verify the identity: \$$(lhs_expr) = $(rhs_expr)\$. Enter 'verified' when complete.",
            answer="verified",
            difficulty=(1850, 2100),
            answer_type="text",
            calculator="scientific",
            solution=steps(
                "Apply the appropriate double angle or sum formula",
                "Simplify both sides to confirm they are equal",
                "The identity is verified"
            ),
            time=180
        )
        
    elseif problem_type == 2
        # Simplify complex trig expression using sum/difference formulas
        a = choice([30, 45, 60])
        b = choice([15, 30, 45])
        
        operation = choice(["+", "-"])
        
        trig_func = choice(["\\sin", "\\cos", "\\tan"])
        
        if trig_func == "\\sin"
            if operation == "+"
                answer_angle = a + b
                formula = "\\sin(A+B) = \\sin A \\cos B + \\cos A \\sin B"
            else
                answer_angle = a - b
                formula = "\\sin(A-B) = \\sin A \\cos B - \\cos A \\sin B"
            end
        elseif trig_func == "\\cos"
            if operation == "+"
                answer_angle = a + b
                formula = "\\cos(A+B) = \\cos A \\cos B - \\sin A \\sin B"
            else
                answer_angle = a - b
                formula = "\\cos(A-B) = \\cos A \\cos B + \\sin A \\sin B"
            end
        else
            if operation == "+"
                answer_angle = a + b
                formula = "\\tan(A+B) = \\frac{\\tan A + \\tan B}{1 - \\tan A \\tan B}"
            else
                answer_angle = a - b
                formula = "\\tan(A-B) = \\frac{\\tan A - \\tan B}{1 + \\tan A \\tan B}"
            end
        end
        
        answer = "$(trig_func)($(answer_angle)°)"
        
        problem(
            question="Simplify \$$(trig_func)($(a)^\\circ $(operation) $(b)^\\circ)\$ to the form \$$(trig_func)(\\theta^\\circ)\$ where \$\\theta\$ is a single angle. Enter your answer as $(trig_func)(angle°).",
            answer=answer,
            difficulty=(1900, 2200),
            answer_type="text",
            solution=steps(
                "Use the formula: \$$(formula)\$",
                "Substitute \$A = $(a)^\\circ\$ and \$B = $(b)^\\circ\$",
                "Result: \$$(answer)\$"
            ),
            time=150
        )
        
    elseif problem_type == 3
        # Product-to-sum or sum-to-product conversion
        conversion_type = rand(1:2)
        
        if conversion_type == 1
            # Product to sum: sin(A)cos(B) = [sin(A+B) + sin(A-B)]/2
            a = choice([3, 4, 5, 6, 7]) * 15
            b = choice([1, 2, 3]) * 15
            
            sum_angle = a + b
            diff_angle = a - b
            
            answer_num = sum_angle + diff_angle
            answer_den = 2
            
            problem(
                question="Use a product-to-sum formula to express \$\\sin($(a)^\\circ)\\cos($(b)^\\circ)\$ as a sum or difference of sines and/or cosines. What is the sum of the two angles (in degrees) that appear in your final answer?",
                answer=answer_num,
                difficulty=(2000, 2300),
                solution=steps(
                    "Apply: \$\\sin A \\cos B = \\frac{1}{2}[\\sin(A+B) + \\sin(A-B)]\$",
                    "Substitute: \$\\sin($(a)^\\circ)\\cos($(b)^\\circ) = \\frac{1}{2}[\\sin($(sum_angle)^\\circ) + \\sin($(diff_angle)^\\circ)]\$",
                    "Sum of angles: \$$(sum_angle) + $(diff_angle) = $(answer_num)\$"
                ),
                time=200
            )
        else
            # Sum to product: sin(A) + sin(B) = 2sin((A+B)/2)cos((A-B)/2)
            a = choice([60, 75, 90, 105, 120])
            b = choice([15, 30, 45, 60])
            
            avg_angle = div(a + b, 2)
            half_diff = div(a - b, 2)
            
            answer = avg_angle
            
            problem(
                question="Use a sum-to-product formula to express \$\\sin($(a)^\\circ) + \\sin($(b)^\\circ)\$ as a product. What is the argument (in degrees) of the sine term in the factored form?",
                answer=answer,
                difficulty=(2050, 2350),
                solution=steps(
                    "Apply: \$\\sin A + \\sin B = 2\\sin\\left(\\frac{A+B}{2}\\right)\\cos\\left(\\frac{A-B}{2}\\right)\$",
                    "Calculate: \$\\frac{$(a)+$(b)}{2} = $(avg_angle)^\\circ\$ and \$\\frac{$(a)-$(b)}{2} = $(half_diff)^\\circ\$",
                    "The sine argument is \$$(avg_angle)^\\circ\$"
                ),
                time=180
            )
        end
        
    elseif problem_type == 4
        # Half-angle formula application
        angle = choice([90, 120, 150, 180, 210, 240])
        half = div(angle, 2)
        
        func_choice = choice(["sin", "cos", "tan"])
        
        if func_choice == "sin"
            formula = "\\sin\\left(\\frac{\\theta}{2}\\right) = \\pm\\sqrt{\\frac{1-\\cos\\theta}{2}}"
        elseif func_choice == "cos"
            formula = "\\cos\\left(\\frac{\\theta}{2}\\right) = \\pm\\sqrt{\\frac{1+\\cos\\theta}{2}}"
        else
            formula = "\\tan\\left(\\frac{\\theta}{2}\\right) = \\frac{\\sin\\theta}{1+\\cos\\theta}"
        end
        
        answer_expr = "$(func_choice)($(half)°)"
        other_func = func_choice == "tan" ? "sin" : "cos"

        problem(
            question="Use a half-angle formula to express \$\\$(func_choice)($(half)^\\circ)\$ in terms of \$\\$(other_func)($(angle)^\\circ)\$. Enter your final simplified answer in the form $(func_choice)(angle°).",
            answer="$(func_choice)($(half)°)",
            difficulty=(1950, 2250),
            answer_type="text",
            calculator="scientific",
            solution=steps(
                "Use the half-angle formula: \$$(formula)\$",
                "Substitute \$\\theta = $(angle)^\\circ\$",
                "Simplify to get \$$(answer_expr)\$"
            ),
            time=210
        )
        
    else
        # Complex identity simplification
        a = nonzero(-5, 5)
        b = nonzero(-5, 5)
        
        # Create an identity like: (sin²x + cos²x) + a·sin(2x) = 1 + a·sin(2x)
        # Answer is the coefficient of the remaining term
        
        identity_type = rand(1:3)
        
        if identity_type == 1
            # sin²x + cos²x = 1 based
            coeff = nonzero(-8, 8)
            answer = abs(coeff)
            
            problem(
                question="Simplify the expression \$$(coeff)(\\sin^2 x + \\cos^2 x) - \\sin^2 x - \\cos^2 x\$ completely. What is the absolute value of the resulting constant?",
                answer=answer,
                difficulty=(1800, 2000),
                solution=steps(
                    "Recognize that \$\\sin^2 x + \\cos^2 x = 1\$",
                    "Substitute: \$$(coeff)(1) - 1 = $(coeff - 1)\$",
                    "Absolute value: \$|$(coeff - 1)| = $(answer)\$"
                ),
                time=120
            )
        elseif identity_type == 2
            # 1 + tan²x = sec²x based
            k = nonzero(-6, 6)
            answer = abs(k)
            
            problem(
                question="Given that \$1 + \\tan^2 x = \\sec^2 x\$, simplify \$$(k)\\sec^2 x - $(k)\\tan^2 x\$ completely. What is the absolute value of the result?",
                answer=answer,
                difficulty=(1850, 2100),
                solution=steps(
                    "Factor: \$$(k)(\\sec^2 x - \\tan^2 x)\$",
                    "Use identity \$\\sec^2 x - \\tan^2 x = 1\$",
                    "Result: \$$(k) \\cdot 1 = $(k)\$, absolute value is \$$(answer)\$"
                ),
                time=150
            )
        else
            # Double angle with coefficient
            m = nonzero(-7, 7)
            target = 2 * m
            
            problem(
                question="If \$\\sin(2x) = 2\\sin x \\cos x\$, and we have \$$(m)\\sin x \\cos x\$, what coefficient \$k\$ satisfies \$$(m)\\sin x \\cos x = k\\sin(2x)\$?",
                answer=m // 2,
                difficulty=(1900, 2150),
                solution=steps(
                    "Use the double angle formula: \$\\sin(2x) = 2\\sin x \\cos x\$",
                    "Therefore: \$\\sin x \\cos x = \\frac{1}{2}\\sin(2x)\$",
                    "So \$$(m)\\sin x \\cos x = \\frac{$(m)}{2}\\sin(2x)\$, giving \$k = $(m//2)\$"
                ),
                time=180
            )
        end
    end
end